src/realm/dictionary.cpp

```diff
@@ -508,17 +617,52 @@ std::pair<Dictionary::Iterator, bool> Dictionary::insert(Mixed key, Mixed value)
     }
 
     auto hash = key.hash();
-    ObjKey k(int64_t(hash & 0x7FFFFFFFFFFFFFFF));
+    ObjKey k(int64_t(hash & s_hash_mask));
 
     bool old_entry = false;
-    ClusterNode::State state;
-    try {
-        // We assume that we will most likely insert new values, so we try this first
+    ClusterNode::State state = m_clusters->try_get(k);
+    if (state.index == realm::npos) {
+        // key does not already exist
         state = m_clusters->insert(k, key, value);
     }
-    catch (const KeyAlreadyUsed&) {
-        state = m_clusters->get(k);
-        old_entry = true;
+    else {
+        // Check if key matches
+        if (do_get_key(state).compare_signed(key) == 0) {
+            old_entry = true;
+        }
+        else {
+            // We have a collision
+            m_clusters->create_collision_column();
+
+            // Create a random key and link it
+            ObjKey random_key;
+            do {
+                // We generate a key with upper bit set. Will not collide with any hash values
+                random_key = ObjKey(int64_t(generator() & s_hash_mask)).get_unresolved();
```

Comment:
I had a comment on this in a previous commit, but it looks like it got lost after a forced push. Is there a reason we're randomly generating keys here? Why don't we just increment the hash value here to get to the next available slot?

Reply:
Ok - I will drop the randomness